File: gospider/parser/parser.go
package parser

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/antchfx/htmlquery"
	"github.com/tidwall/gjson"
	"golang.org/x/net/html"
)

// HTMLParser HTML 解析器
type HTMLParser struct {
	doc  *goquery.Document
	node *html.Node
	html string
}

// NewHTMLParser 创建解析器
func NewHTMLParser(html string) *HTMLParser {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return nil
	}
	node, err := htmlquery.Parse(strings.NewReader(html))
	if err != nil {
		return nil
	}
	return &HTMLParser{
		doc:  doc,
		node: node,
		html: html,
	}
}

// CSS 使用 CSS 选择器提取
func (p *HTMLParser) CSS(selector string) []string {
	results := make([]string, 0)
	query, mode, attr := normalizeCSSSelector(selector)
	if p == nil || p.doc == nil || query == "" {
		return results
	}
	p.doc.Find(query).Each(func(i int, s *goquery.Selection) {
		if value := selectionValue(s, mode, attr); value != "" {
			results = append(results, value)
		}
	})
	return results
}

// CSSFirst 获取第一个匹配
func (p *HTMLParser) CSSFirst(selector string) string {
	results := p.CSS(selector)
	if len(results) > 0 {
		return results[0]
	}
	return ""
}

// CSSAttr 获取属性
func (p *HTMLParser) CSSAttr(selector, attr string) []string {
	results := make([]string, 0)
	query, _, pseudoAttr := normalizeCSSSelector(selector)
	if pseudoAttr != "" {
		attr = pseudoAttr
	}
	if p == nil || p.doc == nil || query == "" || strings.TrimSpace(attr) == "" {
		return results
	}
	p.doc.Find(query).Each(func(i int, s *goquery.Selection) {
		if val, exists := s.Attr(attr); exists && strings.TrimSpace(val) != "" {
			results = append(results, strings.TrimSpace(val))
		}
	})
	return results
}

// CSSAttrFirst 获取第一个属性
func (p *HTMLParser) CSSAttrFirst(selector, attr string) string {
	results := p.CSSAttr(selector, attr)
	if len(results) > 0 {
		return results[0]
	}
	return ""
}

// Links 获取所有链接
func (p *HTMLParser) Links() []string {
	return p.CSSAttr("a", "href")
}

// Images 获取所有图片
func (p *HTMLParser) Images() []string {
	return p.CSSAttr("img", "src")
}

// Title 获取标题
func (p *HTMLParser) Title() string {
	return p.doc.Find("title").Text()
}

// Text 获取文本
func (p *HTMLParser) Text() string {
	return p.doc.Text()
}

// XPathFirst XPath 查询（兼容旧接口）
func (p *HTMLParser) XPathFirst(xpath string) string {
	result, _ := p.XPathFirstStrict(xpath)
	return result
}

// XPath returns every value matched by a full XPath expression.
func (p *HTMLParser) XPath(xpath string) []string {
	values, _ := p.XPathStrict(xpath)
	return values
}

// XPathStrict returns every value matched by a full XPath expression.
func (p *HTMLParser) XPathStrict(xpath string) ([]string, error) {
	if p == nil || p.node == nil {
		return nil, fmt.Errorf("html parser is not initialized")
	}
	if strings.TrimSpace(xpath) == "" {
		return nil, fmt.Errorf("xpath expression is empty")
	}
	nodes, err := htmlquery.QueryAll(p.node, xpath)
	if err != nil {
		return nil, fmt.Errorf("xpath evaluation error: %w", err)
	}
	values := make([]string, 0, len(nodes))
	for _, node := range nodes {
		if value := xpathNodeValue(node); value != "" {
			values = append(values, value)
		}
	}
	return values, nil
}

// XPathFirstStrict runs a full XPath query against the parsed HTML document.
func (p *HTMLParser) XPathFirstStrict(xpath string) (string, error) {
	values, err := p.XPathStrict(xpath)
	if err != nil {
		return "", err
	}
	if len(values) == 0 {
		return "", nil
	}
	return values[0], nil
}

// MustCompileRegex compiles a regex pattern and returns nil on failure.
func MustCompileRegex(expr string) *regexp.Regexp {
	compiled, err := regexp.Compile(expr)
	if err != nil {
		return nil
	}
	return compiled
}

func normalizeCSSSelector(selector string) (query string, mode string, attr string) {
	query = strings.TrimSpace(selector)
	mode = "text"
	lower := strings.ToLower(query)
	if strings.HasSuffix(lower, "::text") {
		return strings.TrimSpace(query[:len(query)-len("::text")]), "text", ""
	}
	if strings.HasSuffix(lower, "::html") {
		return strings.TrimSpace(query[:len(query)-len("::html")]), "html", ""
	}
	attrRe := regexp.MustCompile(`(?i)::attr\(([^)]+)\)\s*$`)
	if match := attrRe.FindStringSubmatchIndex(query); match != nil {
		return strings.TrimSpace(query[:match[0]]), "attr", strings.TrimSpace(query[match[2]:match[3]])
	}
	return query, mode, attr
}

func selectionValue(s *goquery.Selection, mode string, attr string) string {
	switch mode {
	case "attr":
		if value, ok := s.Attr(attr); ok {
			return strings.TrimSpace(value)
		}
	case "html":
		if value, err := s.Html(); err == nil {
			return strings.TrimSpace(value)
		}
	default:
		return strings.TrimSpace(s.Text())
	}
	return ""
}

func xpathNodeValue(node *html.Node) string {
	if node == nil {
		return ""
	}
	if value := strings.TrimSpace(htmlquery.InnerText(node)); value != "" {
		return value
	}
	for _, attr := range node.Attr {
		if strings.TrimSpace(attr.Val) != "" {
			return strings.TrimSpace(attr.Val)
		}
	}
	if value := strings.TrimSpace(htmlquery.SelectAttr(node, "href")); value != "" {
		return value
	}
	return strings.TrimSpace(node.Data)
}

// HTML 获取 HTML
func (p *HTMLParser) HTML() string {
	return p.html
}

// JSONParser JSON 解析器
type JSONParser struct {
	json string
}

// NewJSONParser 创建 JSON 解析器
func NewJSONParser(json string) *JSONParser {
	return &JSONParser{json: json}
}

// Get 获取 JSON 路径
func (p *JSONParser) Get(path string) gjson.Result {
	return gjson.Get(p.json, path)
}

// GetString 获取字符串
func (p *JSONParser) GetString(path string) string {
	return gjson.Get(p.json, path).String()
}

// GetInt 获取整数
func (p *JSONParser) GetInt(path string) int64 {
	return gjson.Get(p.json, path).Int()
}

// GetFloat 获取浮点数
func (p *JSONParser) GetFloat(path string) float64 {
	return gjson.Get(p.json, path).Float()
}

// GetBool 获取布尔值
func (p *JSONParser) GetBool(path string) bool {
	return gjson.Get(p.json, path).Bool()
}

// GetArray 获取数组
func (p *JSONParser) GetArray(path string) []gjson.Result {
	result := gjson.Get(p.json, path)
	results := make([]gjson.Result, 0)
	result.ForEach(func(_, value gjson.Result) bool {
		results = append(results, value)
		return true
	})
	return results
}

File: pyspider/parser/parser.py
"""
HTML 和 JSON 解析器
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from html import unescape
from html.parser import HTMLParser as StdHTMLParser
from typing import Any, Dict, List, Optional, Tuple

try:
    from bs4 import BeautifulSoup  # type: ignore
except ModuleNotFoundError:  # pragma: no cover - exercised only on minimal envs
    BeautifulSoup = None

try:
    from lxml import html as lxml_html  # type: ignore
except ModuleNotFoundError:  # pragma: no cover - exercised only on minimal envs
    lxml_html = None


@dataclass
class _MiniNode:
    tag: str
    attrs: Dict[str, str] = field(default_factory=dict)
    children: List["_MiniNode"] = field(default_factory=list)
    text_chunks: List[str] = field(default_factory=list)

    def text_content(self) -> str:
        parts = list(self.text_chunks)
        for child in self.children:
            parts.append(child.text_content())
        return unescape(" ".join(part for part in parts if part).strip())


class _MiniHTMLTreeBuilder(StdHTMLParser):
    _VOID_TAGS = {"meta", "img", "br", "hr", "input", "link"}

    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.root = _MiniNode(tag="[document]")
        self._stack = [self.root]

    def handle_starttag(self, tag: str, attrs) -> None:
        node = _MiniNode(tag=tag.lower(), attrs={k: v or "" for k, v in attrs})
        self._stack[-1].children.append(node)
        if node.tag not in self._VOID_TAGS:
            self._stack.append(node)

    def handle_startendtag(self, tag: str, attrs) -> None:
        node = _MiniNode(tag=tag.lower(), attrs={k: v or "" for k, v in attrs})
        self._stack[-1].children.append(node)

    def handle_endtag(self, tag: str) -> None:
        normalized = tag.lower()
        while len(self._stack) > 1:
            node = self._stack.pop()
            if node.tag == normalized:
                break

    def handle_data(self, data: str) -> None:
        text = data.strip()
        if text:
            self._stack[-1].text_chunks.append(text)


def _iter_nodes(node: _MiniNode):
    for child in node.children:
        yield child
        yield from _iter_nodes(child)


def _split_selectors(selector: str) -> List[str]:
    return [part.strip() for part in selector.split(",") if part.strip()]


def _parse_simple_selector(selector: str) -> dict:
    pattern = re.compile(
        r"^(?P<tag>[a-zA-Z][\w-]*)?"
        r"(?P<id>#[\w-]+)?"
        r"(?P<classes>(?:\.[\w-]+)*)"
        r"(?:\[(?P<attr>[\w:-]+)(?:=[\"']?(?P<value>[^\"'\]]+)[\"']?)?\])?$"
    )
    match = pattern.fullmatch(selector)
    if not match:
        return {"raw": selector}

    classes = [part for part in match.group("classes").split(".") if part]
    return {
        "tag": (match.group("tag") or "").lower(),
        "id": (match.group("id") or "")[1:],
        "classes": classes,
        "attr": match.group("attr") or "",
        "value": match.group("value") or "",
        "raw": selector,
    }


def _matches(node: _MiniNode, selector: str) -> bool:
    spec = _parse_simple_selector(selector)
    if spec.get("raw") == selector and not any(
        [spec.get("tag"), spec.get("id"), spec.get("classes"), spec.get("attr")]
    ):
        return False

    tag = spec.get("tag")
    if tag and node.tag != tag:
        return False

    node_id = node.attrs.get("id", "")
    if spec.get("id") and node_id != spec["id"]:
        return False

    classes = node.attrs.get("class", "").split()
    for name in spec.get("classes", []):
        if name not in classes:
            return False

    attr = spec.get("attr")
    if attr:
        if attr not in node.attrs:
            return False
        value = spec.get("value")
        if value and node.attrs.get(attr) != value:
            return False

    return True


class HTMLParser:
    """HTML 解析器"""

    def __init__(self, html: str):
        self.html = html
        self._soup = BeautifulSoup(html, "html.parser") if BeautifulSoup else None
        if self._soup is None:
            builder = _MiniHTMLTreeBuilder()
            builder.feed(html)
            builder.close()
            self._root = builder.root
        else:
            self._root = None

    def _fallback_nodes(self, selector: str) -> List[_MiniNode]:
        selectors = _split_selectors(selector)
        nodes: List[_MiniNode] = []
        seen = set()
        for simple_selector in selectors:
            for node in _iter_nodes(self._root):
                if _matches(node, simple_selector):
                    identity = id(node)
                    if identity not in seen:
                        seen.add(identity)
                        nodes.append(node)
        return nodes

    def css(self, selector: str) -> List[str]:
        """CSS 选择器提取"""
        selector, mode, attr = _normalize_css_selector(selector)
        if self._soup is not None:
            return [
                _beautiful_soup_value(elem, mode, attr)
                for elem in self._soup.select(selector)
                if _beautiful_soup_value(elem, mode, attr) != ""
            ]
        return [
            _mini_node_value(node, mode, attr)
            for node in self._fallback_nodes(selector)
            if _mini_node_value(node, mode, attr) != ""
        ]

    def css_first(self, selector: str) -> Optional[str]:
        """获取第一个匹配"""
        selector, mode, attr = _normalize_css_selector(selector)
        if self._soup is not None:
            elem = self._soup.select_one(selector)
            return _beautiful_soup_value(elem, mode, attr) if elem else None
        nodes = self._fallback_nodes(selector)
        return _mini_node_value(nodes[0], mode, attr) if nodes else None

    def css_attr(self, selector: str, attr: str) -> List[str]:
        """获取属性"""
        selector, _, pseudo_attr = _normalize_css_selector(selector)
        attr = pseudo_attr or attr
        if self._soup is not None:
            return [elem.get(attr, "") for elem in self._soup.select(selector)]
        return [node.attrs.get(attr, "") for node in self._fallback_nodes(selector)]

    def css_attr_first(self, selector: str, attr: str) -> Optional[str]:
        """获取第一个属性"""
        selector, _, pseudo_attr = _normalize_css_selector(selector)
        attr = pseudo_attr or attr
        if self._soup is not None:
            elem = self._soup.select_one(selector)
            return elem.get(attr) if elem else None
        nodes = self._fallback_nodes(selector)
        return nodes[0].attrs.get(attr) if nodes else None

    def xpath(self, expr: str) -> List[str]:
        """XPath 提取"""
        if not expr:
            return []

        if lxml_html is not None:
            try:
                tree = lxml_html.fromstring(self.html)
                results = tree.xpath(expr)
                values: List[str] = []
                for item in results:
                    if isinstance(item, str):
                        values.append(item.strip())
                    elif hasattr(item, "text_content"):
                        values.append(item.text_content().strip())
                    else:
                        values.append(str(item).strip())
                return [value for value in values if value]
            except Exception:
                pass

        translated = _translate_simple_xpath(expr)
        if translated is None:
            return []

        selector, attr = translated
        if attr:
            return [value for value in self.css_attr(selector, attr) if value]
        return [value for value in self.css(selector) if value]

    def xpath_first(self, expr: str) -> Optional[str]:
        """获取第一个 XPath 结果"""
        values = self.xpath(expr)
        return values[0] if values else None

    def links(self) -> List[str]:
        """获取所有链接"""
        return self.css_attr("a", "href")

    def images(self) -> List[str]:
        """获取所有图片"""
        return self.css_attr("img", "src")

    def title(self) -> Optional[str]:
        """获取标题"""
        if self._soup is not None:
            title_tag = self._soup.find("title")
            return title_tag.get_text(strip=True) if title_tag else None
        return self.css_first("title")

    def text(self) -> str:
        """获取文本"""
        if self._soup is not None:
            return self._soup.get_text(strip=True)
        return self._root.text_content()


def _translate_simple_xpath(expr: str) -> Optional[tuple[str, Optional[str]]]:
    normalized = expr.strip()
    if not normalized.startswith("//"):
        return None

    attr_match = re.fullmatch(
        r"//([a-zA-Z][\w-]*)(?:\[@([\w:-]+)='([^']+)'\])?/@([\w:-]+)", normalized
    )
    if attr_match:
        tag, filter_attr, filter_value, target_attr = attr_match.groups()
        selector = tag
        if filter_attr and filter_value:
            selector += f"[{filter_attr}='{filter_value}']"
        return selector, target_attr

    text_match = re.fullmatch(
        r"//([a-zA-Z][\w-]*)(?:\[@([\w:-]+)='([^']+)'\])?/text\(\)", normalized
    )
    if text_match:
        tag, filter_attr, filter_value = text_match.groups()
        selector = tag
        if filter_attr and filter_value:
            selector += f"[{filter_attr}='{filter_value}']"
        return selector, None

    class_match = re.fullmatch(r"//([a-zA-Z][\w-]*)\[@class='([^']+)'\]", normalized)
    if class_match:
        tag, class_name = class_match.groups()
        selector = f"{tag}.{class_name.replace(' ', '.')}"
        return selector, None

    attr_filter_match = re.fullmatch(
        r"//([a-zA-Z][\w-]*)\[@([\w:-]+)='([^']+)'\]", normalized
    )
    if attr_filter_match:
        tag, attr_name, value = attr_filter_match.groups()
        selector = f"{tag}[{attr_name}='{value}']"
        return selector, None

    tag_match = re.fullmatch(r"//([a-zA-Z][\w-]*)", normalized)
    if tag_match:
        return tag_match.group(1), None

    return None


def _normalize_css_selector(selector: str) -> Tuple[str, str, Optional[str]]:
    normalized = (selector or "").strip()
    attr_match = re.search(r"::attr\(([^)]+)\)\s*$", normalized, re.IGNORECASE)
    if attr_match:
        return normalized[: attr_match.start()].strip(), "attr", attr_match.group(1).strip()
    for suffix, mode in (("::text", "text"), ("::html", "html")):
        if normalized.lower().endswith(suffix):
            return normalized[: -len(suffix)].strip(), mode, None
    return normalized, "text", None


def _beautiful_soup_value(elem: Any, mode: str, attr: Optional[str]) -> str:
    if mode == "attr" and attr:
        return str(elem.get(attr, "")).strip()
    if mode == "html":
        return "".join(str(child) for child in elem.contents).strip()
    return elem.get_text(strip=True)


def _mini_node_value(node: _MiniNode, mode: str, attr: Optional[str]) -> str:
    if mode == "attr" and attr:
        return node.attrs.get(attr, "").strip()
    return node.text_content().strip()


class JSONParser:
    """JSON 解析器"""

    def __init__(self, json_str: str):
        self.data = json.loads(json_str)

    def get(self, path: str) -> Any:
        """获取 JSON 路径"""
        keys = path.split(".")
        result = self.data
        for key in keys:
            if isinstance(result, dict):
                result = result.get(key)
            elif isinstance(result, list) and key.isdigit():
                result = result[int(key)]
            else:
                return None
        return result

    def get_string(self, path: str) -> Optional[str]:
        """获取字符串"""
        val = self.get(path)
        return str(val) if val is not None else None

    def get_int(self, path: str) -> Optional[int]:
        """获取整数"""
        val = self.get(path)
        return int(val) if val is not None else None

    def get_float(self, path: str) -> Optional[float]:
        """获取浮点数"""
        val = self.get(path)
        return float(val) if val is not None else None

    def get_bool(self, path: str) -> Optional[bool]:
        """获取布尔值"""
        val = self.get(path)
        return bool(val) if val is not None else None

    def get_list(self, path: str) -> Optional[List]:
        """获取列表"""
        val = self.get(path)
        return val if isinstance(val, list) else None
